### PR TITLE
storage/adapter: Introduce the Source References internal table, durably stored in the catalog

### PR DESCRIPTION
**Pull request by rjobanp**
### Motivation

Implements durable storage of 'source references' which will are used to populate the `mz_internal.mz_source_references` introspection table described [here](https://github.com/rjobanp/materialize/blob/subsource-deprecation-blue-green/doc/developer/design/20240625_source_versioning__table_from_sources.md#exposing-available-upstream-references-to-users) as part of the Source Versioning project: https://github.com/MaterializeInc/materialize/pull/27907

This doesn't actually implement the catalog operations to populate this table anywhere, but introduces all the boilerplate necessary to store this data in the catalog durably.

This is part of the work towards https://github.com/MaterializeInc/materialize/issues/28697

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

File: src/catalog/src/builtin.rs [hunk 1]
```diff
@@ -3314,6 +3314,30 @@ pub static MZ_COMMENTS: LazyLock<BuiltinTable> = LazyLock::new(|| BuiltinTable {
...
+            "updated_at",
+            ScalarType::TimestampTz { precision: None }.nullable(false),
+        )
+        .with_column(
+            "columns",
+            ScalarType::List {
+                element_type: Box::new(ScalarType::String),
+                custom_id: None,
+            }
+            .nullable(true),
```
Comment (benesch) [inline src/catalog/src/builtin.rs:3335]: This probably shouldn't actually be null, right? If there are no columns it should be `[]` not `NULL`.

Comment (benesch) [reply src/catalog/src/builtin.rs:3335]: Ah brilliant, thanks! Sorry for the onise.

File: src/catalog/src/durable/upgrade.rs [hunk 2]
```diff
@@ -40,11 +40,14 @@
...
-//! 9. Call your upgrade function in [`run_upgrade()`].
-//! 10. Generate a test file for the new version:
+//! 9. Add an import for your new module to this file: mod v<CATALOG_VERSION-1>_to_v<CATALOG_VERSION>;
+//! 10. Call your upgrade function in [`run_upgrade()`].
+//! 11. Generate a test file for the new version:
 //!     ```ignore
 //!     cargo test --package mz-catalog --lib durable::upgrade::tests::generate_missing_encodings -- --ignored
 //!     ```
+//!     You'll need to update the hashes in `src/catalog/protos/hashes.json` based on the results
+//!     of that command, and add a new one for the new version.
```
Comment (jkosh44) [inline src/catalog/src/durable/upgrade.rs:50]: This is actually not specific to this command but a result of any time you try and build materialize after changing one of the proto files. Can you add this as a separate step after step 5?

File: src/environmentd/tests/testdata/http/ws
Comment (jkosh44) [inline src/environmentd/tests/testdata/http/ws:1]: What happened here?

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/environmentd/tests/testdata/http/ws
Comment (rjobanp) [reply src/environmentd/tests/testdata/http/ws:1]: The system id for the `mz_sources` table changed from `714` to `715` so I had to update the two lines that included the system id in their responses

**Pull request merged**